[RFC] Add Exn module

nojb opened the pull request:
This PR introduces modules `Exn` and `Stack_trace`, corresponding to the built-in type `exn`, taking advantage at the same time to reorganize the `Printexc` API (which would eventually be deprecated).

The differences between `Printexc` and `Exn`/`Stack_trace` are roughly the following:
 
- A type alias `Exn.t` for `exn` is added;
- Functions are renamed/reorganized, as per the table below.

| `Printexc`  | `Exn` / `Stack_trace` |
| ------------- | ------------- |
| `print`  | *removed*  |
| `catch`  | *removed*  |
| `print_backtrace` | *removed* |
| `get_backtrace` | *removed* |
| `record_backtrace` | `Exn.set_tracing` |
| `backtrace_status` | `Backtrace.tracing` |
| `register_printer` |  `Exn.register_printer` |
| `raw_backtrace` | `Stack_trace.t` | 
| `get_raw_backtrace` | `Exn.last_trace` |
| `print_raw_backtrace` | *removed* | 
| `raw_backtrace_to_string` | `Stack_trace.to_string` |
| `raise_with_backtrace` | `Exn.raise_with_trace` | 
| `get_callstack` | `Stack_trace.current` |
| `backtrace_slot` | *removed* |
| `backtrace_slots` | `Stack_trace.slots` |
| `Slot.t` | `Stack_trace.Frame.t` | 
| `Slot.is_raise` | `Stack_trace.Frame.is_raise` | 
| `Slot.is_inline` | `Stack_trace.Frame.is_inline` | 
| `Slot.format` | `Stack_trace.Frame.format` | 
| `Slot.location` | `Stack_trace.Frame.location` | 
| `location` | `Stack_trace.Frame.location` |
| `raw_backtrace_slot` | `Stack_trace.Frame.Raw.t` |
| `raw_backtrace_length` | `Stack_trace.length` | 
| `get_raw_backtrace_slot` | `Stack_trace.Frame.Raw.get` |
| `convert_raw_backtrace_slot` | `Stack_trace.Frame.of_raw` |
| `get_raw_backtrace_next_slot` | `Stack_trace.Frame.Raw.next` |
| `exn_slot_id` | `Exn.id` |
| `exn_slot_name` | `Exn.name` | 

- Lastly, the type `Printexc.t` introduced in #1876 is removed.

nojb committed:
Add Exn module

nojb commented:
Not sure about `print_if_exn`. Personally I never use this function. Maybe it should be moved to `Fun`?

gasche commented:
It's nice to take the time for an API redesign for backtraces (cc @jhjourdan): the previous design was highly constrained by compatibility concerns.

I would have expected to see `raise`, `reraise` and `(re)raise_with_backtrace` also exposed in the new module (cc @bobot). If you are in the mood for giving examples in ocamldoc format, I think that two examples could be useful (and I think writing the examples down could help finalize the API choices):
- catching an exception, printing the backtrace in a warning/alarm and raising a simpler exception
- catching an exception, saving the backtrace, doing some resource handling, and reraising the same exception with the same backtrace.

Finally, if we can bikeshed, I don't like `record_backtrace` as a function name. From the name I expect this function to return or store a backtrace in some way. Maybe `set_backtrace_recording` and `recording_status` would make more sense? Or `record_backtraces`?

nojb commented:
> I would have expected to see raise, reraise and (re)raise_with_backtrace also exposed in the new module (cc @bobot).

Good point, I will add them.

> If you are in the mood for giving examples in ocamldoc format, I think that two examples could be useful (and I think writing the examples down could help finalize the API choices):
>
> - catching an exception, printing the backtrace in a warning/alarm and raising a simpler exception
> - catching an exception, saving the backtrace, doing some resource handling, and reraising the same exception with the same backtrace.

Yes, I will add these. Thanks for the suggestion.

> Finally, if we can bikeshed, I don't like record_backtrace as a function name. From the name I expect this function to return or store a backtrace in some way. Maybe set_backtrace_recording and recording_status would make more sense? Or record_backtraces?

I agree that the names are not good, but couldn't think of obviously better alternatives right away.
We should think of a name that works well for the pair `record_backtrace, backtrace_status`. Maybe `Backtrace.{get,set}_recording` ?

gasche commented:
`get_recording` is not clear to me (I think of a recording as an object that I will get), I think "recording status" (or: configuration, setting, etc.) is the real concept here.

nojb commented:
> get_recording is not clear to me (I think of a recording as an object that I will get), I think "recording status" (or: configuration, setting, etc.) is the real concept here.

OK, what about `Backtrace.{get,set}_recording_status` ?

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,252 @@
+val to_string: t -> string
+(** [to_string e] returns a string representation of the exception [e]. *)
+
+val print_if_exn: ('a -> 'b) -> 'a -> 'b
```
I think this function should simply be dropped. It also has an [error prone](https://github.com/ocaml/ocaml/pull/1855#issuecomment-399756265) signature.

nojb replied:
Yes, I agree.

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,252 @@
+
+    @since 4.08 *)
+
+type t = exn = ..
```
A doc string is missing here.

nojb replied:
Thanks, fixed.

gasche commented:
Yes, I think that would be nice.

alainfrisch commented:
`enable_recording: bool -> unit`, `recording_enabled: unit -> bool`?

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,252 @@
+    standard error output, and the exception is raised again.  The typical use
+    is to catch and report exceptions that escape a function application. *)
+
+val print_current_backtrace: out_channel -> unit
```
As far as the channel function conventions are concerned `print_*` functions are for outputing to `stdout`, `output_*` functions are for functions on arbitrary channel. So it should rather be `output_current_backtrace` --- if at all, see below.

Personally I don't see a good reason to have `Exn` depend on the channel abstraction I would rather move to remove these functions altogether, possibly adding them to the future `Channel` module with the right names and provide a good API that allows us to easily format backtraces

nojb replied:
This function is mostly for convenience; I agree that we could be more orthogonal by removing it (and also `string_of_current_backtrace` below). Personally I wouldn't mind doing so. Are there any opinions against?

gasche replied:
It's fine to remove them. Historically, those functions were the first available ( 640b24249d8e284905f3db44899ffc06e006f028 ), the more atomic operations only came later.

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,252 @@
+
+val string_of_current_backtrace: unit -> string
+(** Returns a string containing the same exception backtrace that {!print}
+    would print. Same restriction usage than {!print}. *)
```
Is that function really needed ? Isn't this `Backtrace.to_string Exn.current_callstack  ()` ?

nojb replied:
No, it is `Backtrace.to_string (Backtrace.current ())`. But I agree that maybe this function could be eliminated.

gasche commented:
This would also be fine with me, although I find @nojb's proposal more systematic (it is easy to guess one name from the other).

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,252 @@
+(** Returns [true] if exception backtraces are currently recorded, [false] if
+    not. *)
+
+module Backtrace : sig
```
TBH I find this design quite messy can't we maybe review the organisation and prune a bit the terminology ?

Do we really need both call stack and backtrace in the vocabulary ? Isn't an "exception backtrace" simply an exception with a call stack ? Wouldn't it make sense to maybe have these `Backtrace`s in their own toplevel module as `Callstack` ? I mean `Exn.current_callstack` really doesn't feel satisfactory to me. 

If that is done then maybe rather than recording backtraces we can simply trace exceptions and maybe have `Exn.{tracing,set_tracing}`.

nojb replied:
I agree `current_callstack` is odd, but where to put it?

Note that while there is only one type here ("backtraces/call stacks") there are two notions of "current": "current backtrace" means the backtrace/call stack of the "last" raise; "current callstack" means the backtrace/call stack at the call site. So in any case we will need some terminology to differentiate these two notions.

As long as backtraces/call stacks remain closely related to exceptions, it doesn't seem strange to me to keep them in a submodule of `Exn` (the issue of `current_callstack` notwithstanding). But if we expect them to have a life of their own (are we?) maybe putting them in their own top-level module could be considered.

dbuenzli replied:
> Note that while there is only one type here ("backtraces/call stacks") there are two notions of "current": "current backtrace" means the backtrace/call stack of the "last" raise; "current callstack" means the backtrace/call stack at the call site. So in any case we will need some terminology to differentiate these two notions.

Well `Exn.current_trace ()` or even `Exn.last_trace ()` and `Callstack.current ()`.

gasche replied:
> As long as backtraces/call stacks remain closely related to exceptions,

Daniel has a point that while backtraces are related to exceptions, call stacks are an independent concept. Currently there is only one primitive type `raw_backtrace`, but we could have a `raw_callstack` type (with the same underlying representation), or a less opinionated name shared between the two, such as "stack trace" or "stack fragment".

(I don't have time to devote to these API questions before mid-November, so unfortunately I won't participate in making proposals of my own. Good luck!)

nojb committed:
Add raise, reraise, raise_notrace

nojb committed:
Rename record_{backtrace,status} => Backtrace.{set,get}_recording_status

nojb committed:
Add docstring

nojb committed:
Remove print_if_exn, print_current_backtrace, string_of_current_backtrace

nojb committed:
Doc

nojb commented:
- Added `raise`, `reraise` and `raise_notrace` as suggested by @gasche, and
- Removed `print_current_backtrace`, `string_of_current_backtrace`, `print_if_exn`, `Exn.print` as suggested by @dbuenzli

dbuenzli commented:
> * Added `raise`, `reraise` and `raise_notrace` as suggested by @gasche, and

So we have `raise_notrace` but `raise_with_backtrace`. Could we maybe choose between `trace` and `backtrace` ? 

I would rather try to go with the former, then the we can call these things "exception traces" and the whole process "exception tracing" which in turn might allow to get rid of the notion of `record`ing (same would work with `backtrace` though "exception backtracing" feels a little bit more odd).

nojb committed:
Add Callstack module

nojb commented:
I went ahead and added a separate `Callstack` module. It does feel like a cleaner design. I also switched from `backtrace` terminology to just `trace` or `exception trace` when needed, so as suggested by @dbuenzli, we have `Exn.tracing: unit -> bool`, `Exn.set_tracing: bool -> unit` and `Exn.last_trace: unit -> Callstack.t`.

nojb committed:
Doc

dbuenzli commented on stdlib/callstack.mli:
```diff
@@ -0,0 +1,139 @@
+    @since 4.08 *)
+
+type t
+(** The abstract type [t] stores a callstack.
```
There's a mix of "callstack" and "call stack" in the text, one should be chosen. Maybe this suggests it should rather be `Call_stack`; stack already appears separately in `Stack_overflow`.

nojb replied:
Personally, I find `Call_stack` a bit ugly. Maybe use `callstack` everywhere? Another possibility is to use `Backtrace` instead of `Callstack` as module name.

dbuenzli replied:
Backtrace is less clear than "call stack" or "stack trace" to me. I don't mind if you want to call it `Callstack` and if you use "call stack" in the text but at least do it consistently. 

That said if people never use "callstack" as a single word, conventions would say that `Call_stack` is the right identifier. 

You could also call it `Stack_trace` which would, terminlogy-wise, blend better with `Stack_overflow` exception.

dbuenzli replied:
In fact maybe  "stack trace" is a better description of the value than "call stack". If I have a value for a call stack I would expect to be able to actually manipulate the call stack with it, which is not the case.

nojb replied:
OK, let's go with `Stack_trace`. I pushed a commit with the renaming.

hannesm commented:
I'm not entirely convinced of the name `Exn` - is there a reason to not name it `Exception`?

dbuenzli commented:
> I'm not entirely convinced of the name `Exn` - is there a reason to not name it `Exception`?

It's the name of the type in Pervasives.

nojb committed:
Rename Callstack to Stack_trace

nojb committed:
Add implementation

nojb committed:
Implement Printexc in terms of Exn and Stack_trace

nojb committed:
make depend

nojb committed:
Update manual

nojb committed:
Changes

nojb committed:
Remove Printexc.t

nojb committed:
Avoid inlined wrapper around external

nojb commented:
I filled out the implementation, updated the manual, and rewrote the implementation of `Printexc` to call the functions from the new modules.

nojb committed:
check-typo fixes

dbuenzli commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+
+val current: int -> t
+(** [current n] returns a description of the stack trace on the current program
+    point (for the current thread), with at most [n] entries. *)
```
That's an annoying API since it forces me to make an arbitrary choice (`n`). I mean in practice it's likely that what I want is "all of them", so my gut usage would be simply to plug `max_int` here. 

Can't we maybe make that argument optional, that is having `current : ?limit:int -> unit -> t` and such that an unspecified `limit` indicates all or as much as possible ?

nojb replied:
Modified as suggested.

nojb replied:
The wrapper around the primitive `caml_get_current_callstack` (the old `Printexc.get_callstack`) appears in the stack trace, which is not nice; shall we modify the primitive itself to take the optional argument?

nojb replied:
I pushed one commit using an optional argument for the primitive directly, seems to do the trick.

dbuenzli commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+(** [length st] returns the number of (non-inlined) slots in the stack trace
+    [st]. *)
+
+module Slot : sig
```
I don't find `Slot` to be a great name. It seems `Java` simply call these [StackTraceElement](https://docs.oracle.com/javase/10/docs/api/java/lang/StackTraceElement.html) and `.NET` [StackFrame](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframe?view=netframework-4.7.2). Maybe `Frame` or `Site` ? I would tend to favour the latter.

nojb replied:
I have a slight preference for `Frame` rather than `Site`, which sounds a bit too generic to me.

dbuenzli replied:
My reasoning behind favouring `Site` over `Frame` is basically the same I had [here](https://github.com/ocaml/ocaml/pull/2137#discussion_r231470042) but since this is in the `Stack_trace` module rather than at toplevel maybe that's fine. Though I would still argue that the information we have here is really about the call and raise *sites* and nothing else (IIUC this is a source location + metadata about the location); for a frame I would expect information about call parameters, local variables, etc.

mshinwell replied:
I would suggest that these are called "frames", which matches standard terminology in e.g. `gdb`, but the function that obtains a location from a frame is called "`call_site`" or similar.

dbuenzli replied:
I trust @mshinwell to have the right name.

nojb replied:
OK, I changed the name of the module to `Frame` and the function `Stack_trace.slots` to `Stack_trace.frames`.

dbuenzli commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+    val next: t -> t option
+    (** [next slot] returns the next inlined slot, if any.
+
+        Sample code to iterate over all frames (inlined and non-inlined):
```
Note that "frames" is used here instead of "slots".

nojb replied:
Thanks, fixed the docs to use "frame(s)" everywhere.

dbuenzli commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+    { filename: string;
+      line_number: int;
+      start_char: int;
+      end_char: int }
```
I wouldn't expose the record fields but rather move it to its own module with a creating function and accessors.

mshinwell replied:
Indeed.  Also, I would recommend putting the closing curly brace on its own line (as you have in other places), since it makes it quicker to add another record field.

nojb replied:
Do we want to expose a "create" function, or just accessors?

dbuenzli replied:
I think it would definitively be useful be able to create values or redefine fields e.g. to remap filenames in a given context.  I would also rename `{start,end}_char` to `{start,end}_byte` since it's likely a better characterization of what it is but that may not be consistent with the rest of the compiler apis.

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,100 @@
+    If [fn] raises an exception, both the exceptions passed to [fn] and raised
+    by [fn] will be printed with their respective stack traces. *)
+
+(** {1 Exception slots} *)
```
I don't understand this heading. Maybe these two functions could simply be moved before or after the `to_string` function.

nojb replied:
Indeed, I moved them to the very top, before the `to_string` function.

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,100 @@
+
+val to_string: t -> string
+(** [to_string e] returns a string representation of the exception [e]. *)
+
```
I think a heading `{1:tracing Tracing}` could be introduced for functions related to tracing.

nojb replied:
Thanks, done.

dbuenzli commented on stdlib/exn.mli:
```diff
@@ -0,0 +1,100 @@
+    handler), the stack trace may correspond to a later exception than the
+    handled one. *)
+
+external raise: t -> 'a = "%raise"
```
These functions are missing a doc string. I would move them before the tracing one and under their own `{1:raising Raising}` heading.

nojb replied:
I added some doc and put them under their own heading.

mshinwell commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+    In the return array, the slot at index [0] corresponds to the most recent
+    function call or raise.
+
+    Some possible reasons for returning [None] are as follow:
```
"as follow" -> "as follows"

nojb replied:
Thanks, fixed.

mshinwell commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+
+val length: t -> int
+(** [length st] returns the number of (non-inlined) slots in the stack trace
+    [st]. *)
```
Why would this only return the number of non-inlined frames?  Seems a bit odd.

nojb replied:
As far as I can tell, this reflects the runtime representation of stack traces, so it is more efficient than "flattening" the stack traces fully. I wonder if this cost is worth the complication in the API.

mshinwell commented on stdlib/stack_trace.mli:
```diff
@@ -0,0 +1,141 @@
+
+val slots: t -> Slot.t array option
+(** Returns the slots of a stack trace, or [None] if none of them contain useful
+    information.
```
The presence of this option type seems a bit strange.  Maybe it would be better to always return the slots, and then have accessor functions that demonstrate what properties they have?

nojb replied:
For one thing, the frames (slots) may not be obtainable if no debug info is available.

dbuenzli replied:
Do we want to distinguish between no debug info available and an empty array though ?

nojb replied:
I don't see why one would want to.

nojb replied:
OK, I threw away the option and and now return an empty array instead of `None`.

nojb committed:
Rename Stack_trace.{Slot => Frame, slots => frames}

nojb committed:
Reorder the interface of Exn, add doc headings

nojb committed:
Add doc for raise functions

nojb committed:
Typo

nojb committed:
Use optional argument for Stack_trace.current

nojb committed:
caml_get_current_callstack: add optional argument variant

nojb committed:
Stack_trace.frames: return an empty array instead of None

nojb commented:
I am thinking of getting rid of the notion of "raw" stack trace frames (type `Printexc.raw_backtrace_slot` or `Stack_trace.Frame.Raw.t`) and just exposing the "friendly" type (type `Printexc.backtrace_slot` or `Stack_trace.Frame.t`).

What do you think?

gasche commented:
Some use cases for callstack slots are performance-sensitive. For example, @jhjourdan uses them in the statistical profiler (you want to collect them for a reasonably high portion of your allocations while keeping a reasonably low overhead), and I would expect that they are traversed in reporting/statistics/aggregation logic which may also traverse non-small amount of data and have to be fast.

I haven't followed the recent discussions (slowly coming back to the real world^W^Wpull requests), and don't have a full picture of the API (maybe having a low-level notion of trace, but no low-level notion of slot, could be enough), but we designed the original API exposure with the idea that easy things should be easy and advanced/low-level things possible, and I liked that design choice.
(On the other hand, it certainly poses some tensions, for example in term of API evolution with evolving runtime features; inlined slots came only after, and disrupted some ideas we had about linear slot access. cc @let-def)

let-def commented:
I think @diml also has some performance sensitive use cases for callstacks.
I would prefer to keep the low-level API available, making it explicit that the compatibility might be hard to preserve, rather than removing them.

Raw traces have to be kept as they are the ones that can be reinstalled with `%raise_with_backtrace` and afaict, the low-level API does not add much complications (if we ignore backward compatibility) while it is useful to implement the high-level API internally.

That being said, I would like the normal API to be as simple as possible, the `length` function for instance could stop distinguishing between regular and inlined functions. (... be should the stack trace type reflect the "physical" state of the stack or the "logical" one, that includes the inlined functions?).

let-def commented:
... I might also add a richer version of stack traces which includes the names and values of locals in the future.

ghost commented:
I did some experiment at some point, but right now we have nothing in production that is performance sensitive and relies on callstacks.

Currently, the only operations you can do on a `raw_backtrace_slot` is convert it and get the next entry. This allows to iterate over a raw backtrace until a certain point without allocating `backtrace_slot` values for the whole backtrace. It'd be nice to preserve this feature and it seems like it could be done without the `raw_backtrace_slot` type.

gasche commented:
I heard feedback recently (during the MirageOS retreat) that the exception-backtrace API is hard to use, so I still think that there would be user interest in having an `Exn` module. If someone else than @nojb is willing to give this a try, it could end up being a very helpful change -- but it certainly will be somewhat frustrating in the middle.
